### ISSUE TITLE
How many fish version should we support?

### ISSUE THREAD

**Issue by edouard-lopez** [rafaelrinaldi/pure#123]
As I'm updating the [_.travis.yml_ config to run our tests against multiple versions](https://github.com/rafaelrinaldi/pure/pull/122), I'm wondering how many `fish` versions should we support?

Currently [TravisCI build](https://travis-ci.org/rafaelrinaldi/pure/builds/477606183):
* :heavy_check_mark: succeed against `2.5.0`, `2.6.0`, `2.7.1`, 
* :red_circle:  fail against `2.4.0` and `3.0.0`.

What support strategy should we adopt here?
@rafaelrinaldi @schrodincat @jorgebucaran @faho

**Comment by jorgebucaran** [comment]
Fish >=2.3 would be nice. 2.3 has conf.d and the `string` builtin—what else do we need? :)

Fisher supports +2.1 right now, but I hope to make it >=2.3 too by next year.

**Comment by schrodincat** [comment]
I’d say we support only 2.7 and 3. I see no reason to stay on older versions from a user point of view, since there is no backwards incompatible changes between minor updates.

**Comment by faho** [comment]
I'm always looking at supported Ubuntu releases - https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=fish:

>
    trusty (14.04LTS): 2.0.0-1
    xenial (16.04LTS): 2.2.0-3
    bionic (18.04LTS): 2.7.1-3
    cosmic (18.10): 2.7.1-3build1

Trusty is only supported for another 3 months, and all of these are supported by [the official PPA](https://launchpad.net/~fish-shell/+archive/ubuntu/release-3) (with 3.0!). 19.04 will have 3.0.

But in the end it depends on how much pain it causes to support older releases - anything < 2.3.0 is rough because those lack `string`, which means you'd have to fall back to `sed` and `grep`, at which point you're almost dealing with another language. Snippets also were only introduced in 2.3.0.

In contrast, 2.4.0 didn't have as many big changes when it comes to scripting - it added the `realpath` builtin and changed how `^` redirections were handled. It did add `math -s` (but not the short `--scale`).

2.5.0 added the subcommand-style to `status` (`status is-interactive` instead of `status --is-interactive`), though both remain supported. It also adds `command -sq`.

2.6.0 added `disown` and fixed `(fzf)`.

2.7.0 added `argparse`, allowed invalid array indices, a few nice additions to `string`, setting $COLUMNS and $LINES correctly.

3.0.0 added better `set -lx`, a new `math` builtin, a rather important fix to index ranges (`$var[2..-1]` expanded to $var[1] if i didn't have 2 elements in 2.7.0, and errored out before), "logical" $PWD, the path variable changes, `&&` et al, the `wait` builtin, `bind --preset`, ....

-----

So, the biggest pain point is probably 2.2.0. Anything before that, you can safely ignore, but the difference between what you'd write in 2.2.0 and what you'd write in 2.3.0 is massive.

2.4.0 is probably most important because of `math -s` - otherwise you'd have to truncate numbers to integers manually for compatibilitiy with 3.0, if you need that. (i.e. `math -s0 2 / 3` or `math 2 / 3 | string replace -r '\..*' ''`)

2.5.0's `command -sq` is nice, but not all that necessary.

2.6.0 doesn't have much interesting for us, 2.7.0's argparse is big, 3.0.0 is _massive_.

**Comment by faho** [comment]
Okay, the tests on 2.4.0 fail for two reasons:

- It didn't have `fish_default_mode_prompt`

- It didn't have `status $subcommand`, so `status something` prints output before it fails

3.0 fails because of $hostname and the fish_title isn't completely correct.

I think the following should work for 2.4.0's status:

```fish
    set --local current_command (status current-command 2>/dev/null)
    or set --local current_command $_
```

and you'd have to figure something out for the default mode prompt.

**Comment by edouard-lopez** [comment]
### :heavy_check_mark: `2.5`, `2.6`, `2.7.1` and `3.0.0` 
Tests successfully run [against `2.5`, `2.6`, `2.7.1` and `3.0.0` on CI](https://travis-ci.org/rafaelrinaldi/pure) :sunglasses: 

### :red_circle:  `2.4` 
Supporting `2.4` should be possible, as [only 4 tests fail](https://travis-ci.org/rafaelrinaldi/pure/jobs/478610445):
```
# tests 122
# pass  118
# fail  4
```
However, I [removed it from the CI configuration](https://github.com/rafaelrinaldi/pure/pull/128/commits/bdb87102640ddc423f393da54172db025898b878) bdb871026 cause the trade-off time invested/benefit isn't worth it from my point of view as I want to focus on other projects.

### :pray: Contribute
If you want to investigate there is a [makefile](https://github.com/rafaelrinaldi/pure/blob/master/makefile) that can help you get started and run tests in `docker` containers:

```
❯ make
usage:  make build-pure-on FISH_VERSION=3.0.0   # build container
        make test-pure-on  FISH_VERSION=3.0.0   # run tests
        make dev-pure-on   FISH_VERSION=3.0.0   # dev in container
```

You might be interested by the [helpers to condition tests execution](https://github.com/rafaelrinaldi/pure/blob/master/tests/pure_tools_versions-compare.test.fish), based on the `$FISH_VERSION` (cf. 
[_pure_prompt_ssh_host.test.fish](https://github.com/rafaelrinaldi/pure/blob/833a8e76e725de308f1d46191e3a5872f781e320/tests/_pure_prompt_ssh_host.test.fish#L32-L40).

### :rocket:  Docker

Running in a `docker` image require to adapt some of our code, hence the use of some short form option on some executable (_e.g._ `sort`, `hostname`, etc. See #127). Still prefer long form if possible.

**Comment by edouard-lopez** [comment]
Adding a message at session start-up to notify the user he should upgrade would be a nice feature!

Created issue #131 for that purpose.

**Comment by imbsky** [comment]
It seems like Pure already supports fish 3. Isn't it supported yet?

**Comment by edouard-lopez** [comment]
@imbsky we do support fish `3.0.0`. I define [_supported_](https://github.com/rafaelrinaldi/pure/blob/master/.travis.yml#L33-L38) as _test passing against said version_

**Comment by imbsky** [comment]
@edouard-lopez I thought so... [This warning](https://github.com/rafaelrinaldi/pure/blob/master/README.md#install) made me very confused. So I just made a PR.

**Comment by edouard-lopez** [comment]
Based on @faho idea, I would like to restrict our support to versions available, by default, and above, in supported Ubuntu LTS and current Ubuntu release only. So, as of Nov. 15 2020 we have:

* ~xenial (16.04LTS): `2.2.0-3`~ we won't go back to this as we decided to support above `≥2.5`
* bionic (18.04LTS): `2.7.1-3`
* focal (20.04LTS): `3.1.0-1.2`
* groovy (shells): `3.1.2-1ubuntu1`

Henceforth we should:
* :x:  drop support for ~`2.5`~, ~`2.6`~ officially
* :heavy_check_mark: set lowest supported version to `2.7.1`
* :tada: introduce support for `3.1.0` and above

What do you think?

**Comment by jorgebucaran** [comment]
I think we should support >=`3.0`, otherwise we'll be stuck in `2.7.1` until 2023 for no substantial benefit. Bionic/Cosmic users can upgrade to the latest Fish easily enough.

**Comment by jorgebucaran** [comment]
Support >=`3.0`, otherwise we'll be stuck in `2.7.1` until 2023 for no substantial benefit. Bionic/Cosmic users can upgrade to the latest Fish easily enough.

**Comment by jorgebucaran** [comment]
@smorimoto Could you create a different issue for that? Not sure what's wrong with that picture, though.

@edouard-lopez Sounds good! 🎉

**Comment by smorimoto** [comment]
@jorgebucaran Of course. Ah, it was a little hard to understand with the image.
![gif](https://user-images.githubusercontent.com/38746192/99329052-b88cd000-28c0-11eb-9ab3-1ca189e2201d.gif)

**Comment by jorgebucaran** [comment]
@smorimoto Maybe you have a `fish_greeting.fish` file somewhere doing something weird? Look in your $__fish_config_dir.

**Comment by smorimoto** [comment]
@jorgebucaran I'm specifying an empty for that; is that a problem? ​This behavior did not occur before.

**Comment by jorgebucaran** [comment]
@smorimoto Nope, that shouldn't be the issue AFAIK. This seems like a Fish specific issue, so I'd suggest asking in https://gitter.im/fish-shell/fish-shell.

**Comment by smorimoto** [comment]
The default prompt for fish doesn't cause this, and it happened after updating fisher and pure.

**Comment by edouard-lopez** [comment]
I'm closing the issue as we have a [new support strategy](https://github.com/rafaelrinaldi/pure/issues/123#issuecomment-728189321).

@smorimoto Please open a dedicated issue so we have a clear thread.

**Comment by edouard-lopez** [comment]
[Versions Support Strategy](https://github.com/rafaelrinaldi/pure/wiki/Versions-Support-Strategy) has been added to the wiki for reference
